### DISCUSSION TITLE
working-with-stacked-branches-in-git-is-easier-with-update-refs/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#75] [Announcements]
# working-with-stacked-branches-in-git-is-easier-with-update-refs/

In this post I discuss how to use the new Git rebasing feature, --update-refs, and how it makes working with stacked branches/PRs easier.

https://andrewlock.net/working-with-stacked-branches-in-git-is-easier-with-update-refs/

**Comment by david-bakin-sl** [comment]
Excellent post - subject and description of it - and just in time for me as I've just started working with stacked PRs - and boy oh boy is `--update-refs` going to be a lot easier than doing it manually!  Passing this post around where I work.

**Comment by fleutot** [comment]
I understand that this is to show a use of `--update-refs`, but I don't see the advantage of stacked PR. If I have a large enough feature implementation, which I can divide into parts to ease reviews and history clarity, I usually create a feature branch. I use that feature branch like you normally would use a `dev` branch, but only for that feature. 

PR for all "sub-parts" are based on the feature branch, and merge to the feature branch. You get small PR, the parts are showing in the history, and it also shows that they were part of the feature branch. What this lacks, I guess, is the dependencies, such as "You must merge Part4 before Part5", but stacked PR have that problem as well?

The merge commit texts "Merge feature/x.y.z/part3 into feature/x.y.z/part2" does not really make sense, does it? The parts are not necessarily related, a merge commit that says "Merge feature/x.y.z/buttons into feature/x.y.z/rest_api" seems awkward. Better have both parts merge commits say "Merge feature/x.y.z/buttons into feature/x.y.z", and "Merge feature/x.y.z/rest_api into feature/x.y.z".

With a feature branch, you also get to possibility to review the whole feature implementation (the sum of all parts), when time has come to merge it to `dev`.

**Reply by alkeryn** [reply]
it can be very useful if for example you want to remove/merge an old commit from all branch without having them now being completely separate.

**Reply by fleutot** [reply]
Remove an old commit, as in rewrite history? This is frowned upon anyway, isn't it? There's the use case where someone committed a password in clear text, but how often does this happen...

**Reply by david-bakin-sl** [reply]
> Remove an old commit, as in rewrite history? This is frowned upon anyway, isn't it? There's the use case where someone committed a password in clear text, but how often does this happen...

"this is frowned upon anyway" is a simplification.  There are good use cases for removing old commits.  There's no difficulty with it in a local copy of a repo - otherwise we wouldn't encourage squashing before pushing in order to make things clearer.  There's no difficulty in a remote copy of a repo either in any case where noone is already sharing that code - e.g., while the code is still sequestered in an unaccepted PR.  And ... this entire post is about working with PRs.

**Reply by alkeryn** [reply]
yea or just take for example, i'm working on some project that i plan to open source but it is still closed source and i'm the only contributor, it has many branches, and i want to clean it up a little before publishing it.

for example i forgot i had a commit called "tmp" in the history and i merged it with all branches.
i also renamed the project once.

rebases can be really useful for that kind of things and --update-refs just makes life easier.

**Reply by jermainedilao** [reply]
It's common if you're on trunk based development. Merging a huge `feature` branch onto `dev` branch can be sometimes quite a headache.

**Reply by mcepl** [reply]
It is very useful for all those `devel` branches: you know when you post [plenty of submissions upstream](https://lists.sr.ht/~martanne/devel/patches?search=status%3Aproposed+from%3A%7Emcepl), and now you are waiting on upstream to react, e.g. https://git.sr.ht/~mcepl/vis/log/devel .

**Comment by david-bakin-sl** [comment]
Well I just tried this and discovered that the "middle" branches of my 5-deep stacked PR didn't get updated.  All my branches are already pushed of course - I was going to force push when done - but when I did for example

```
git checkout pr.5
git rebase --update-refs pr.1
```

The only tags that got updated were pr.1 and pr.5 - pr.{2,3,4} were unchanged.

Difference between what I did and what you're showing is I didn't use any form of "qualified" name for the branches.  e.g. you were using `andrew/...`.

What should I have been doing?  naming the remotes?  naming the local (how?)

**Comment by yndajas** [comment]
Spotted a typo: `Git version 23.8` -> `Git version 2.38`

**Comment by pvspain** [comment]
Hi Andrew, great post - thank you!

This is a great feature, which deserves wider use. But it does invite the usual kvetching about _rebase versus merge_, as you alluded.

Would you consider writing an article about the git workflow for a coworker on the receiving end of all those forced pushes? I'm assuming `--update-refs` can come to the rescue there as well? Maybe its just another use-case to append to this post?

Demonstrating a relatively painless way to deal with forced pushes on the receiving end may take some of the heat out of the _rebase versus merge_ debate?      

Cheers, Paul.

**Reply by jmlane** [reply]
Agreed. Force pushing to shared branches is where this methodology falls apart for me.

**Reply by mcepl** [reply]
Well, the `rebase`-v-`merge` thing. I think the absolute rule about never rebasing applies only to the **public** branches (e.g., `master` or maintenance branches), but this whole stacked branches thing is meant mostly for your internal development branches, and I don’t know about you, but I rebase these constantly. If some of your colleagues working with you on a new feature wants to follow such stacked branch (or one or more of those stacked branches more likely), then, yes, you would have to tell them to `git reset --hard jmlane/feature` or `git rebase jmlane/feature` after every `git fetch`.

**Comment by inkarkat** [comment]
One use case where enabling `--update-refs` by default is problematic is when rebasing _submodules_ in a project that uses a superproject repo to tie several modules together.

The automatic update of the stacked branches in a submodule would sever the reference in the corresponding superproject branch (Git does not and in fact cannot update those references in a separate repo). So we unfortunately still have to go through each stacked branch in the superproject and commit / amend the (updated) submodule references manually.

But this is just a minor caveat for an otherwise helpful feature; thanks for the excellent presentation!

**Comment by collinstevens** [comment]
I face a similar issue a lot; but this doesn't quite work for me. In your example, the PR Feedback commit is reordered to part-1, but I would have squashed it into commit 3. That results in a new hash and rebasing the dependent branches breaks.

Is there a way to handle this or am I still stuck in my cherry-pick workflow?

**Reply by andrewlock** [reply]
There is, I was planning a post describing my normal workflow, but tl;dr; --onto with git rebase solves most of these issues!

**Reply by collinstevens** [reply]
> There is, I was planning a post describing my normal workflow, but tl;dr; --onto with git rebase solves most of these issues!

Does your workflow also handle when the part-1 PR additionally is squash merged into master? I assume --onto also solves that somewhat?

Is it just an interactive rebase, but dropping the part-1 commit? Even then, I don't know how that would fix the other branches because they wouldn't also drop the part-1 commit I assume.

**Reply by collinstevens** [reply]
Hmm, after playing around with stacked PRs and --update-refs more, it's just cursed and bait.

For example, if you want to rebase latest master changes, you are SOL to do so. I do this multiple times throughout the day.

Or if your company is squash merging PRs in, this means the commit ID changes so now you have to checkout all your branches and drop the merged commit and rebase each of them.

I experimented with this workflow because I have 8 "stacked PRs" at a client at the moment and it was becoming a PITA to work with.

I've gone back to my workflow where I essentially have an integration branch and then 8 cherry-picked branches which all have 8 active PRs for. If I want to pull master in, I rebase the integration branch and then I recreate the cherry-picked branches. I delete all of them, create them off of master, and then cherry-pick each commit from the integration branch over.

In my normal case, my commits aren't stacked as in they are dependent on the previous, but I prefer to work with all my latest changes together. If my commits were dependent, I would follow the same workflow, but the cherry-picks would be a range instead of a single commit ID.

It seems I need an --update-refs flag for this type of workflow, but which works with cherry-picks and also knows about squash merging. I imagine it would be possible to script this passing it the commit IDs which should be dropped and then it would be able to fix all the branches.

**Reply by andrewlock** [reply]
You're describing the workflow I have, and while it's a little annoying, it's not that hard to deal with 🤔 

Lets say you have 5 PRs stacked on `master`
```
master --- PR1 --- PR2 --- PR3 --- PR4 --- PR5
```

You then merge PR1 to master, so it's a bit like this:

```
---  master 
 \--- PR1 --- PR2 --- PR3 --- PR4 --- PR5
```

I would handle it one of two ways, either:

```bash
git checkout PR5
git rebase PR1 --onto origin/master
git branch -D PR1
```

_or_

```bash
git checkout PR5
git rebase -i origin/master
# this opens the interactive branch, delete all the PR1 commits and the ref to it
git branch -D PR1
```

If I was just on the command line, I'd do option 1, but I'm often in Rider, in which case I can do option 2 from the UI with shortcuts.

And it just works for me 🙂

**Reply by collinstevens** [reply]
Hmm, I wasn't aware when you rebase PR5 and drop the commits, it would also drop the commits from PR2/PR3/PR4.

Okay, this will work for me then.

**Reply by andrewlock** [reply]
> Hmm, I wasn't aware when you rebase PR5 and drop the commits, it would also drop the commits from PR2/PR3/PR4

I guess I wasn't clear, this _wouldn't_ drop the commits from PR2/PR3/PR4. 😅 

If you start with this:

```
master --- PR1 --- PR2 --- PR3 --- PR4 --- PR5
```

and then merge PR1 to master, so it's a bit like this:

```
---  masterV2 
 \--- PR1 --- PR2 --- PR3 --- PR4 --- PR5
```

then if you run 

```bash
git checkout PR5
git rebase PR1 --onto origin/master
git branch -D PR1
```

then you end up with this:


```
masterV2 --- PR2 --- PR3 --- PR4 --- PR5
```

Is that not what you want?

**Reply by collinstevens** [reply]
Hmm, this is the outcome I desire, but I don't think this works when commit `PR1` is squashed into master. `PR1` becomes `PR1'`, a new commit ID.

Maybe --onto does something different than I thought and handles this?

This is why I said I would have to go drop `PR1` in interactive rebases from all the branches when I rebased them off of `master`.

I'm going to test this out.

**Reply by collinstevens** [reply]
Okay, I tested this over here https://github.com/collinstevens/stacked-pr-tests with branches `master`, `part-1`, `part-2`, and `part-3`; definitely works.

I was under the impression it wouldn't be able to handle the different commit IDs.

 - in branch `part-1`, the commit `part-1 commit` ID is `f4e9481`.
 - in branch `master`, the commit `part-1 commit` ID is `4b9bddc` after the squash merge into `master`.

I'll have to go read exactly how --onto works...

**Reply by aaccioly** [reply]
@andrewlock, I appreciate the great article and comments. Many thanks.

@collinstevens, it works because `git rebase` defaults to `--empty=drop`, i.e., it drops empty commits.

From [Git's documentation for `git-rebase`](https://git-scm.com/docs/git-rebase/2.45.0):

> The current branch is reset to `<upstream>` or `<newbase>` if the `--onto` option was supplied. This has the exact same effect as `git reset --hard <upstream>` (or `<newbase>`). `ORIG_HEAD` is set to point at the tip of the branch before the reset.

Meaning that for all practical purposes, you can think of `git rebase PR1 --onto origin/main` (or similarly of `git switch part-1 && git rebase origin/main`) as a command which starts with:

```sh
git reset --hard origin/main
```

I.e., in your example, `rebase` will start from `main` which contains all changes from `PR1` (given that you have just squashed `part-1` changes into `main`).

Also from the documentation:

> If the upstream branch already contains a change you have made (e.g., because you mailed a patch which was applied upstream), then that commit will be skipped and warnings will be issued (if the **merge** backend is used). For example, running `git rebase master` on the following history (in which `A'` and `A` introduce the same set of changes, but have different committer information):
> ```
>          A---B---C topic
>        /
>    D---E---A'---F master
> ```
> will result in:
> ```
>                  B'---C' topic
>                  /
>    D---E---A'---F master
> ```

Which ties nicely to the `--empty` option:

> **`--empty=(drop|keep|stop)`**
> How to handle commits that are not empty to start and are not clean cherry-picks of any upstream commit, but which become empty after rebasing (because they contain a subset of already upstream changes):
>
> **`drop`**
> The commit will be dropped. This is the default behavior.

Once established that all commits from `part-1` will be dropped as part of the rebase (meaning that `part-1` will point to the same commit as `main`), from here on, `--update-refs` will work its magic, and `part-2` will be transplanted to `main`, and so on and so forth.

---

*TL;DR:* In the scenario where `PR1` was just squashed into `main`, rebasing `PR1` onto `main` implies that all changes from `PR1` are empty and will be dropped by default.

**Reply by mcepl** [reply]
> Hmm, this is the outcome I desire, but I don't think this works when commit `PR1` is squashed into master. `PR1` becomes `PR1'`, a new commit ID.

Or to be precise, you get a merge conflict and when you resolve it (preferring changes already on `main`, obviously) and `git rebase --continue`, the already merged commit gets dropped.

**Comment by mcepl** [comment]
`git rebase --update-refs` is really cool (I didn’t know it could work together with `git rebase -i` and most likely with `git commit --fixup= && git rebase --autosquash`), but now I would like to have also `git push <all-those-branches-from-dev-excluded>`. Any ideas?

**Reply by alkeryn** [reply]
not sure if there is a git way to do it but you could always do something like git branch -a | <your filters using the shell | xargs git push

**Reply by dfaure-kdab** [reply]
Actually a colleague of mine and me came up with a zsh function and alias to do exactly that....
Haven't used it much yet, though, so use with care....

function gpb() {
  UPSTREAM="$(git remote | grep origin || git remote | grep gerrit)"
  GERRIT="$(git remote | grep gerrit)"
  BRANCH="${$(git symbolic-ref -q --short "refs/remotes/$UPSTREAM/HEAD"):-"$UPSTREAM/master"}"
  git rev-list --ancestry-path "$(git merge-base HEAD "$BRANCH")"..HEAD | xargs git name-rev --refs='refs/heads/*' --name-only | sed '/~[0-9]\+$/d' | while read r; do
     if [[ $r != wip-* ]]; then
         if [ -n "$GERRIT" ]; then
             git push gerrit ${r}:refs/for/master
         else
             git push $* origin "$r"
         fi
     fi
  done
}
alias git-push-all='gpb --force-with-lease --force-if-includes'

**Reply by dfaure-kdab** [reply]
Hmm, formatting got messed up I think, and I can't edit, here it is as a download: http://www.davidfaure.fr/2024/push-all

**Reply by mcepl** [reply]
If we are talking shell script, is this what you meant?

```sh
#!/bin/sh

gpb() {
  UPSTREAM="$(git remote | grep origin || git remote | grep gerrit)"
  GERRIT="$(git remote | grep gerrit)"
  MYHEAD="$(git symbolic-ref -q --short "refs/remotes/$UPSTREAM/HEAD")"
  BRANCH="${MYHEAD:-"$UPSTREAM/master"}"
  git rev-list --ancestry-path "$(git merge-base HEAD "$BRANCH")"..HEAD \
  | xargs git name-rev --refs='refs/heads/*' --name-only | sed '/~[0-9]\+$/d' \
  | while read -r r; do
     case "$r" in
     wip-*)
         if [ -n "$GERRIT" ]; then
             git push gerrit "${r}":refs/for/master
         else
             git push "$@" origin "$r"
         fi
         ;;
     *)
         ;;
     esac
  done
}
alias git-push-all='gpb --force-with-lease --force-if-includes'
```

**Reply by dfaure-kdab** [reply]
Sounds right (my version used zsh syntax). Does it work for you? ;)

**Reply by dfaure-kdab** [reply]
ah no, the other way around. wip-* is my naming for branches that shouldn't be pushed. You want to push in the `*)` case.

**Reply by bzhb** [reply]
I have a more concise way to do it:
```sh
git config --global alias.stack-branches "log --decorate-refs=refs/heads --decorate=short --simplify-by-decoration --pretty=%D ^origin/main HEAD"
```

Then typing `git stack-branches` will return the list of the stack branches (granted that your reference branch is `main` and that your stack is on top of it).

To push all of them:
```sh
git stack-branches | xargs -I {} git push --force-with-lease origin {}
```

That can be put in a shell alias, for me as I'm using fish:
```fish
alias -s git-stack-push="git stack-branches | xargs -I {} git push --force-with-lease origin {}"
```

**Reply by mcepl** [reply]
Doesn't work well with repos, where the trunk branch is not called `main`. What about this?

```sh
stack-branches log --decorate-refs=refs/heads --simplify-by-decoration --pretty=%D ^origin/HEAD HEAD
```

**Reply by bzhb** [reply]
@mcepl You are right, it is better with `origin/HEAD`, thanks !

Also there is an issue with `--pretty=%D` when there is several branches on the same commit (the branch then appears on the same line separated with a comma), so I changed it to `--pretty=format:"%(decorate:prefix=,suffix=,tag=,separator=%n)"`

The updated version:
```
git config --global alias.stack-branches "log --decorate-refs=refs/heads --simplify-by-decoration --pretty=format:\"%(decorate:prefix=,suffix=,tag=,separator=%n)\" ^origin/HEAD HEAD"
```

**Reply by miguelpduarte** [reply]
If all the upstreams are set for each branch I'm pretty sure this can be simply:

```
git push --force-with-lease origin branch-1 branch-2 branch-3 ...
```

And `origin` of course could be something else, but is most often just `origin`.

Docs: https://git-scm.com/docs/git-push#Documentation/git-push.txt-refspec

**Reply by ktaragorn** [reply]
I use this on windows, and couldnt get `xargs` to play ball..but 

`git push --force-with-lease origin $(git stack-branches) `

Seemed to work well.

**Reply by miguelpduarte** [reply]
> I use this on windows, and couldnt get `xargs` to play ball..but
> 
> `git push --force-with-lease origin $(git stack-branches) `
> 
> Seemed to work well.

@ktaragorn just make sure to quote the shell substitution as otherwise special characters in branch names can cause issues. You can run it through shellcheck to confirm 🙂

**Reply by ktaragorn** [reply]
Thanks, good to know!

**Comment by cadolphs** [comment]
This is really cool and comes in quite handy. Just one scenario isn't covered in the article yet: What happens if you squash-merge part 1? Then, from all I can tell, you actually _do_ need some `git --rebase --onto main part1 part3 --update-refs` or so. 

Or maybe I'm missing something?

**Reply by andrewlock** [reply]
Yep, you got it right, you need to rebase the full "remaining" stack in that case

**Reply by dfaure-kdab** [reply]
But if you keep a clean local history (where your fixup commits are squashed together already), then a simple rebase without arguments (after a fetch, obviously) will automatically skip the already merged commits. That's very nice.
